### ISSUE TITLE
Console php:// streams - fopen failed to open: Too many open files

### ISSUE THREAD

**Issue by thumbtech** [cakephp/cakephp#12255]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.6.5
* Platform and Target:
Welcome to Ubuntu 16.04.4 LTS (GNU/Linux 4.4.0-1061-aws x86_64)
PHP 7.1.16 (cli) (built: Apr  2 2018 22:39:46) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.1.16, Copyright (c) 1999-2018, by Zend Technologies
bitnami@ip-172-31-25-0:~$

### What you did
In CLI (bin/cake) I am using dereuromark/cakephp-queue and calling dispatchShell in a task, so essentially dispatchShell in a loop with potentially hundreds of repetitions.

### What happened
Process dies after "many" (a somewhat large but unpredictable number) iterations, always with a PHP error to the effect that:
fopen(php://stdin) failed to open stream: Too many open files
[Note sometimes died on php://stderr instead of stdin]
... referencing the expected fopen lines in Console/ConsoleInput.php or Console/ConsoleOutput.php

### What you expected to happen
Based on searching for this type of error and eventually noting recommended treatment of these streams discussed here:
http://php.net/manual/en/wrappers.php.php

... I experimented making the following adjustment to Console/ConsoleInput.php contstructor:

```
    public function __construct($handle = 'php://stdin')
    {
        $this->_canReadline = (extension_loaded('readline') && $handle === 'php://stdin');
        switch ($handle)
        {
            case 'php://stdin':
                $this->_input = STDIN;
                break;
            default:
                $this->_input = fopen($handle, 'rb');
        }
    }
```

... and Console/ConsoleOutput.php contstructor:

```
    public function __construct($stream = 'php://stdout')
    {
        switch ($stream)
        {
            case 'php://stdout':
                $this->_output = STDOUT;
                break;
            case 'php://stderr':
                $this->_output = STDERR;
                break;
            default:
                $this->_output = fopen($stream, 'wb');
        }

        if ((DIRECTORY_SEPARATOR === '\\' && !(bool)env('ANSICON') && env('ConEmuANSI') !== 'ON') ||
            (function_exists('posix_isatty') && !posix_isatty($this->_output))
        ) {
            $this->_outputAs = self::PLAIN;
        }
    }
```

FWIW this did eliminate the errors I was seeing (with no apparent downside) and is consistent with the PHP recommendation that one should use the constants instead of opening these streams manually ...

**Comment by markstory** [comment]
Did you do any digging into why the file handles were not closed after each dispatchShell() call? Closing files when they are no longer in use is another way to solve too many open file handles.

**Comment by thumbtech** [comment]
I considered fclosing and did not explore how / where this could be easily accomplished in this scenario. Another option is to just play with ulimits in the OS. It does seem though from the PHP docs that fopen(php://stdout) etc. is a redundant/duplicate handle which nevertheless counts against open file limit, so best to just use STDOUT which is the "default" handle supplied via PHP CLI ... ?

**Comment by markstory** [comment]
ConsoleOutput already has a `__destruct` which should clean up open file handles. However, ConsoleInput does not. Are you able to check if `ConsoleOutput::__destruct()` is ever called?

**Comment by thumbtech** [comment]
I want to point out that I had to revert to a PHP 5.6 install due to an unrelated issue. This also reverted my hacked cake Console code. I have not encountered the "Too many open files" issue even though I would have expected it with similar test runs as before.

This may indicate a different treatment of these streams in PHP 7, but I did not find any indications of this or related bugs.php.net entries, so I'm not sure.

I still return to the fact that the PHP manual explicitly discourages fopen('php://stdout') etc., but am less confident that this is a defect per se.
